### PR TITLE
Use Github REST API and Index Commit Messages off Github Repository

### PR DESCRIPTION
**Pull request by debanjum**
- [X] Migrate to Github REST API instead of Llama Hub to index Markdown Docs in Repository
- [X] Index Commit Messages from Github Repository as well

### CODE REVIEW

File: src/khoj/processor/github/github_to_jsonl.py [hunk 1]
```diff
@@ -64,26 +66,89 @@ def process(self, previous_entries=None):
...
+
+        while commits_url is not None:
+            # Get the next page of commits
+            response = requests.get(commits_url, headers=headers)
+
+            # If the rate limit is reached, wait for the reset time
+            if response.status_code != 200 and response.headers.get("X-RateLimit-Remaining") == "0":
+                wait_time = int(response.headers.get("X-RateLimit-Reset")) - int(time.time())
+                logger.info(f"Github Rate limit reached. Waiting for {wait_time} seconds")
+                time.sleep(wait_time)
```
Comment (sabaimran) [inline src/khoj/processor/github/github_to_jsonl.py:117]: nit: Can move this into a utility function within this class, e.g., `def _check_if_rate_limited_and_wait(response)`.

File: src/khoj/processor/github/github_to_jsonl.py [hunk 2]
```diff
@@ -64,26 +66,89 @@ def process(self, previous_entries=None):
...
+                # Create URL for each markdown file on Github
+                url_path = f'https://github.com/{self.config.repo_owner}/{self.config.repo_name}/blob/{self.config.repo_branch}/{item["path"]}'
+
+                # Get text from each markdown file
+                file_content_url = f'{self.repo_url}/contents/{item["path"]}'
+                headers["Accept"] = "application/vnd.github.v3.raw"
+                markdown_file_contents = requests.get(file_content_url, headers=headers).content.decode("utf-8")
+
+                # Add markdown file contents and URL to list
+                markdown_files += [{"content": markdown_file_contents, "path": url_path}]
```
Comment (sabaimran) [inline src/khoj/processor/github/github_to_jsonl.py:99]: Nice.

**Review by sabaimran — APPROVED** [review]
Overall looks like a great improvement on top of the existing integration 🚀 . My one concern is around how long this would take to retrieve & index the data.

**Comment by sabaimran** [comment]
Did you explore using this [Python Github API library](https://github.com/PyGithub/PyGithub)?

**Comment by debanjum** [comment]
> Overall looks like a great improvement on top of the existing integration 🚀 . My one concern is around how long this would take to retrieve & index the data.

Fair concern on how long this could take. If you don't hit rate-limits, it can get done in 10-20s, from my testing. If you do hit rate limits than you could have to wait a while (maybe 15-30minutes! for the indexing to complete). Some potential optimizations
- Doing the retrieval in an async call in the background
- Asking for commits, changed files since last retrieval can also reduce probability of hitting rate 
- Doing the retrieval once a day (rather than once every hour) can also maybe be tried

These can maybe be added in follow on updates to the plugin

**Comment by debanjum** [comment]
> Did you explore using this [Python Github API library](https://github.com/PyGithub/PyGithub)?

No, i hadn't seen it yet. 
1. They seem to be [actively looking for maintainers](https://github.com/PyGithub/PyGithub#maintainership) so sticking with calling Github API directly maybe better for long-term stability?
2. Their documentation doesn't mention how to get all commits and seems fairly sparse
3. They do provide multiple authentication mechanisms and some examples on how to use webhooks that could be useful

File: src/khoj/processor/github/github_to_jsonl.py [hunk 3]
```diff
@@ -64,26 +66,89 @@ def process(self, previous_entries=None):
...
+        # get the contents of the repository
+        response = requests.get(repo_content_url, headers=headers)
+        contents = response.json()
+
+        # If the rate limit is reached, wait for the reset time
+        if response.status_code != 200 and response.headers.get("X-RateLimit-Remaining") == "0":
+            wait_time = int(response.headers.get("X-RateLimit-Reset")) - int(time.time())
+            logger.info(f"Github Rate limit reached. Waiting for {wait_time} seconds")
+            time.sleep(wait_time)
+            return self.get_markdown_files()
```
[the remark this answers is not in the document]
Comment (debanjum) [reply src/khoj/processor/github/github_to_jsonl.py:84]: Umm, where here? I've already added the rate limiting code here. It's been added to the 2 places I see us calling Github; 1. to get all markdown files and 2. to get all commits in a repository

File: src/khoj/processor/github/github_to_jsonl.py [hunk 4]
```diff
@@ -64,26 +66,89 @@ def process(self, previous_entries=None):
...
+
+                # Add markdown file contents and URL to list
+                markdown_files += [{"content": markdown_file_contents, "path": url_path}]
+
+        return markdown_files
+
+    def get_commits(self) -> List[Dict]:
+        # Get commit messages from the repository using the Github API
+        headers = {"Authorization": f"{self.config.pat_token}"}
+        commits_url = f"{self.repo_url}/commits"
```
[the remark this answers is not in the document]
Comment (debanjum) [reply src/khoj/processor/github/github_to_jsonl.py:106]: No, don't see how. The /git/trees/ endpoint just returns a snapshot of the directory at given point/branch head, so a list of git trees (directories) and blobs (files). No commit history is available via this endpoint from what I understand

File: src/khoj/processor/github/github_to_jsonl.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (debanjum) [reply src/khoj/processor/github/github_to_jsonl.py]: I've made the request to get content of a markdown file wait if it hits rate limiting

**Pull request merged**